File: server_JAVA/src/main/java/com/example/demo/mapper/UserMapper.java
package com.example.demo.mapper;

import org.mapstruct.Mapper;
import org.mapstruct.factory.Mappers;

import com.example.demo.dto.NewUserDTO;
import com.example.demo.models.User;

@Mapper
public interface UserMapper {
    UserMapper INSTANCE = Mappers.getMapper(UserMapper.class);
    
    NewUserDTO toDTO(User user);

    User toEntity(NewUserDTO userDTO);
}

File: server/src/main/kotlin/com/example/demo/user/mapper/UserMapper.kt
package com.example.demo.user.mapper

import com.example.demo.recipe.mapper.RecipeMapper
import com.example.demo.user.domain.User
import com.example.demo.user.domain.dto.UserDTO
import org.springframework.stereotype.Component

@Component
class UserMapper(private val recipeMapper: RecipeMapper) {
    fun toDTO(user: User): UserDTO {
        return UserDTO(
            id = user.id,
            email = user.email,
            username = user.username,
            recipes = user.recipes.map { recipeMapper.toDTO(it) },
        )
    }
}
